consul/kv_consul.go

```diff
@@ -498,18 +498,39 @@ func (kv *consulKV) LockWithID(key string, lockerID string) (
 	*kvdb.KVPair,
 	error,
 ) {
+	return kv.LockWithTimeout(key, lockerID, kvdb.DefaultLockTryDuration, kv.GetLockTimeout())
+}
+
+func (kv *consulKV) LockWithTimeout(
+	key string,
+	lockerID string,
+	lockTryDuration time.Duration,
+	lockHoldDuration time.Duration,
+) (*kvdb.KVPair, error) {
 	key = stripConsecutiveForwardslash(key)
 	// Strip of the leading slash or else consul throws error
 	if key[0] == '/' {
 		key = key[1:]
 	}
 
-	l, err := kv.getLock(key, lockerID, 20*time.Second)
-	if err != nil {
-		return nil, err
-	}
-	if _, err := l.lock.Lock(nil); err != nil {
-		return nil, err
+	timeout := time.After(lockTryDuration)
+	var (
+		l    *consulLock
+		err  error
+		done bool
+	)
+	for !done {
+		select {
+		case <-timeout:
+			return nil, err
```

Comment:
Can you check if err is still nil, then return an actual error for try timeout expired?

Reply:
If someone sets the timeout to a small value and getLock takes more than that, then we might end up returning nil instead of an error. Is this not a valid scenario?

Reply:
get lock is blocking.